### DISCUSSION TITLE
Function that creates multiple streams with only one ERC-20 transfer

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#1114] [Ideas]
We've recently received this feature request from a user:

<img width="400" src="https://github.com/user-attachments/assets/bfa49ac9-7233-4d44-a560-9db930ed23c5"/>

@smol-ninja answered promptly and correctly:

<img width="345" alt="SCR-20241211-rwxb" src="https://github.com/user-attachments/assets/c57e43f0-486b-4853-b1e3-a7c8068b1776" />

But the user made a good point.

Having a `createMultiple` function that performs only one ERC-20 would bring huge benefits.

1. Lower the gas cost for all users.
2. Make it possible to create more than ~60 streams with a single CSV file upload.

I know we're already close to the maximum size allowed for contracts, but maybe there's a way to make this happen somehow?

cc @sablier-labs/solidity

**Comment by PaulRBerg** [comment]
Can I get your thoughts on this @sablier-labs/solidity, specifically if you think the contract size allows this feature to be implemented?

It's extremely important feature because it would significantly increase the number of streams that can be created in one go.

**Comment by smol-ninja** [comment]
### How many additional streams can this feature support?

I conducted some tests to evaluate this. Here's what I did:
1. Copied the `BatchLockup.createWithDurationsLL` logic into `SablierLockup` and called it `createWithDurationsLLMultiple`
2. Moved the token transfer logic from the private `_create` function to `createWithDurationsLLMultiple`, making sure a single transfer of tokens
3. Created 100 streams using `createWithDurationsLLMultiple` as well as current `createWithDurationsLL` and `batch` functions (which includes 100 transfers)

#### Gas Benchmarks
1. `createWithDurationsLLMultiple` with single transfer: 11.8M gas
2. `batch` function (100 transfers): 11.9M gas (0.7% higher than above)
3. Periphery’s `BatchLockup.createWithDurationsLL` (100 transfers): 12.4M gas (5% higher than a single transfer)

The periphery’s batch function incurs the highest cost because it calculates the total transfer amount by summing up the values from each batch entry. This same check was performed in `createWithDurationsLLMultiple`, and that is why the gas reduction is only 5%. If we decide to skip this check and trust that the UI or users will input the correct total amount, we can improve gas efficiency slightly more, potentially reducing it to 7-10%.

Based on these results, my guess is it would increase the stream capacity by 7-10.

---------------

### How much gas can be saved

~5%, and if we do not perform the transfer amount check then probably ~10%.

-------------------

### Why does the UI only allow 60 streams?

it’s possible to create approximately 150 streams in a single transaction based on [gas benchmark](https://docs.sablier.com/guides/lockup/gas-benchmarks).

------------------------

### Contract size

If we modify `createWithDurationsLL` to accept an array (similar to the batch lockup implementation) instead of a single stream input, this change can fit within the current contract size. For a single stream, the array would only contain one entry, and so the gas cost should remain comparable.

**Reply by PaulRBerg** [reply]
Great sleuthing, thanks for looking into this, Shub!!

Boy was I wrong on the magnitude on this one! Only 5-10% 😂

It looks like the lion's share of gas is consumed by the Sablier storage logic, not by the ERC-20 transfer.

Given that the potential benefits are so small, I don't think it would be worth it to implement this. Especially since having separate ERC-20 transfers makes accounting and inspection on Etherscan easier.

> my guess is it would increase the stream capacity by 7-10

LOL

> Why does the UI only allow 60 streams?

Good question: https://github.com/sablier-labs/v2-interfaces/issues/1271

**Reply by smol-ninja** [reply]
> Boy was I wrong on the magnitude on this one! Only 5-10% 😂

Haha. Even I was shocked to see that. A simple ERC20 transfer consumes 31k gas so by that logic, 100 transfers should consume 3.1M gas.

When I disabled ERC20 transfer completely in the Lockup contract, it consumed 11.8M gas to create 100 streams with no token transfer compared to 12.4M with 100 transfers. So in this context, it costed ~0.5M gas to make 100 transfers (~5k per transfer). Thats 4 streams worth of gas.

**Comment by PaulRBerg** [comment]
I wonder how much more gas we would save if instead of performing these calculations for every item in the batch array:

https://github.com/sablier-labs/lockup/blob/e105895c589b5811358257e7ac15fe7605493060/src/SablierLockup.sol#L372-L378

We would perform them only once, like so:

```solidity
nextStreamId = streamId + batch.length;
aggregateBalance[params.token] += totalAmount;
```

**Reply by smol-ninja** [reply]
Will look into it this week.

**Reply by andreivladbrg** [reply]
We can perform a benchmark but I personally I don't find a good pattern to design the core contract so that it is optimized for batching.

**Reply by PaulRBerg** [reply]
If we can enable 10-20 more streams, it makes complete business sense to do it. Aesthetics should take a back seat in the decision calculus.

**Reply by smol-ninja** [reply]
> nextStreamId = streamId + batch.length

If we perform this, we won't be able to return all the stream IDs created by the user. We can, though, return starting stream ID and ending stream ID and expect user to understand that but that might be problematic. So, imo we should not change this logic.

----------

- I refactored code so that there is only one ERC20 transfer.
- Update aggregate amount only once.

### Results:

1. [Tx 1](https://sepolia.etherscan.io/tx/0x4a990a36a5282ead92aab00c37039ff10e2e07754098556ebb1ba203b6c573ad) with refactored code to create 100 Linear streams: consumes 16.2 M gas.
1. [Tx 2](https://sepolia.etherscan.io/tx/0x94ffc7de4a8113f5438604cd4c3e22b06430de63d6cd2878d103eada870b2c0b) with `staging` code to create 100 Linear streams using `batch`: consumes 17.14 M gas.

Thats 5% saving compared to `batch`.

So compared to `BatchLockup`, there could be another 0.5M gas difference. That means we can allow ~7-8 more streams on 16M gas limit. So on a full block, its ~14-16.

### Some notes on refactoring:

The code can be seen [here](https://sepolia.etherscan.io/address/0xeb61501871c22793f58384d83e041dcfb6181e39#code).

#### Function interface

```solidity
 function createWithDurationsLL(
        IERC20 token,
        BatchLockup.CreateWithDurationsLL[] calldata batch
    )
```

Accepting `totalAmount` avoids one extra loop.

#### Single ERC20 transfer

```solidity
// Transfer the total amount to the contract.
token.safeTransferFrom({ from: msg.sender, to: address(this), value: amountDeposited });

aggregateAmount[token] += amountDeposited;
```

`amountDeposited` is summed inside the create loop.

<details><summary>Click to see code</summary>

```solidity
 function createWithDurationsLL(
        IERC20 token,
        BatchLockup.CreateWithDurationsLL[] calldata batch
    )
        external
        payable
        override
        noDelegateCall
        returns (uint256[] memory streamIds)
    {
        // Check that the batch size is not zero.
        uint256 batchSize = batch.length;
        if (batchSize == 0) {
            revert Errors.SablierBatchLockup_BatchSizeZero();
        }

        uint128 amountDeposited;

        // Use the block timestamp as the start time.
        uint40 startTime = uint40(block.timestamp);

        streamIds = new uint256[](batchSize);

        for (uint256 i = 0; i < batchSize; ++i) {
            // Set the current block timestamp as the stream's start time.
            Lockup.Timestamps memory timestamps = Lockup.Timestamps({ start: startTime, end: 0 });

            uint40 cliffTime;

            // Calculate the cliff time and the end time.
            if (batch[i].durations.cliff > 0) {
                cliffTime = timestamps.start + batch[i].durations.cliff;
            }
            timestamps.end = timestamps.start + batch[i].durations.total;

            // Checks, Effects and Interactions: create the stream.
            streamIds[i] = _createLL(
                Lockup.CreateWithTimestamps({
                    sender: batch[i].sender,
                    recipient: batch[i].recipient,
                    depositAmount: batch[i].depositAmount,
                    token: token,
                    cancelable: batch[i].cancelable,
                    transferable: batch[i].transferable,
                    timestamps: timestamps,
                    shape: batch[i].shape
                }),
                batch[i].unlockAmounts,
                cliffTime
            );

            amountDeposited += batch[i].depositAmount;
        }

        // Transfer the total amount to the contract.
        token.safeTransferFrom({ from: msg.sender, to: address(this), value: amountDeposited });

        aggregateAmount[token] += amountDeposited;
    }
```
</details>

---------

Gas difference is not as we expected but we should consider the change as its slightly more optimized and we wont have to manage `BatchLockup` anymore.

**Reply by andreivladbrg** [reply]
Good job on the very in-depth benchmark, Shub — it looks you put in a lot of effort

So the difference between 16.2 and 17.14 is: **5.64%** — still not that big.

*(Note: the `totalAmount` param can be removed and we can simply transfer `amountDeposited`. It’s a small gas improvement, but it won’t make a significant difference.)*

**Reply by PaulRBerg** [reply]
To @smol-ninja: great sleuthing, thanks!

> we can allow ~7-8 more streams on 16M gas limit. So on a full block, its ~14-16.

Those are non-negligible metrics, even more so when considering that Base has 120M blocks, which means ~60 streams more!

> we won't be able to return all the stream IDs created by the user

We can actually derive all stream IDs from the new `nextStreamId` and the batch size.

> as its slightly more optimized and we wont have to manage `BatchLockup` anymore.

Let's do it! We've had users who were confused about `BatchLockup` in the past.

---

To @andreivladbrg:

> So the difference between 16.2 and 17.14 is: 5.64% — still not that big.

Not big, but relevant.

**Reply by smol-ninja** [reply]
> Good job on the very in-depth benchmark

Thanks.

> it looks you put in a lot of effort

Without a grain of doubt.

> the difference between 16.2 and 17.14 is: 5.64% — still not that big

Thats between the new code and the `batch` function. The difference between the new code and `BatchLockup` would be 8-10%. In terms of streams, that is equivalent to creating 7-8 more streams with 16M gas limit.

> Note: the totalAmount param can be removed and we can simply transfer amountDeposited. It’s a small gas improvement, but it won’t make a significant difference.)

Good catch and you are right. We can handle `safeTransferFrom` after the loop using `amountDeposited` (will update it).

**Reply by PaulRBerg** [reply]
> > it looks you put in a lot of effort
>  Without a grain of doubt.

🫡

**Reply by andreivladbrg** [reply]
> Not big, but relevant

<img width="500" alt="image" src="https://github.com/user-attachments/assets/5bc520be-7eee-40ef-b466-e6914a5099e0" />

I agree that’s relevant — but I assumed you still believe it’s not worth it.

**Reply by smol-ninja** [reply]
> We can actually derive all stream IDs from the new nextStreamId and the batch size.

Yes we can but I think @sablier-labs/frontend relies on events for stream IDs (I am assuming you mean to derive it off-chain).

**Reply by PaulRBerg** [reply]
Ooh. Right. In this case, we may still need to emit the ID in each stream.

**Comment by PaulRBerg** [comment]
Related issue:

[Refactor SablieLockup create functions to allow array of parameters](https://github.com/sablier-labs/lockup/issues/1235)

**Comment by smol-ninja** [comment]
Question from https://github.com/sablier-labs/lockup/pull/1251:

What if we only keep create with timestamps and remove create with durations? Do we have data on which is more popular?

**Reply by PaulRBerg** [reply]
We definitely need both.

In Legacy, we used to only have `createWithTimestamps` (it wasn't called like this but you get the point).

There were *a lot* of requests for `createWithDurations`. An especially important example is DAO proposals where `createWithDurations` where they would like to start streaming/vesting only after the proposal has been voted and passed.

**Reply by smol-ninja** [reply]
Yeah OK. Lets keep both. It was just an idea.

**Comment by smol-ninja** [comment]
I propose that we abandon this feature and stick with `SablierLockupBatch`.

While the pros are clear, the biggest disadvantage is the lack of flexibility to add more functionalities to Lockup. After [implementing this feature](https://github.com/sablier-labs/lockup/pull/1251), we are left with only a 28B margin at 300 runs. This makes it an extremely tight contract which would force us to continuously debate whether a new feature can be squeezed in or not. In the medium term, these limitations will likely roll us back to the `SablierLockupBatch` approach. 

The only reason for implementing this feature was to allow users to create more streams within a single tx. But it comes at a huge cost.  We should wait until EIP-7907 (or a similar EIP) is included into the EVM before considering it again.

While I was in favour before, I am strongly against it now until we absolutely want to freeze Lockup to not have any changes at all. The batchification is a very expensive change in terms of contract size. 

@sablier-labs/evm for comments.

**Reply by smol-ninja** [reply]
For examples, its
- https://github.com/sablier-labs/lockup/discussions/1253
- https://github.com/sablier-labs/lockup/issues/1219
- https://github.com/sablier-labs/lockup/issues/663
- More features in the future

vs Batchification.

**Reply by PaulRBerg** [reply]
Thanks for the update @smol-ninja.

The contract size limitation is sooo annoying.

I agree with your conclusion.

**Reply by andreivladbrg** [reply]
same here, agree with the conclusion 

<img width="400" alt="image" src="https://github.com/user-attachments/assets/f702dd85-cd3e-4967-b579-0d09c9d691fa" />

**Reply by smol-ninja** [reply]
Thanks guys. Will close it now.
